File: demo/src/main/java/com/example/java/com/example/JsonlTraceSink.java
package com.example;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;

public class JsonlTraceSink implements TraceSink {

    private BufferedWriter writer;
    private ObjectMapper objectMapper;

    public JsonlTraceSink() {
        this.objectMapper = new ObjectMapper();
        // For making appropriate Date format
        this.objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
        setupFile();
    }

    private void setupFile() {
        try {
            // 1. logs klasörünü oluştur
            File directory = new File("logs");
            if (!directory.exists()) {
                directory.mkdirs();
            }

            // 2. Dosya adını oluştur (run-YYYYMMDD-HHMMSS.jsonl)
            String timeStamp = new SimpleDateFormat("yyyyMMdd-HHmmss").format(new Date());
            String fileName = "logs/run-" + timeStamp + ".jsonl";

            // 3. Writer'ı başlat (append: true)
            this.writer = new BufferedWriter(new FileWriter(fileName, true));
            
            System.out.println("Trace log started: " + fileName);

        } catch (IOException e) {
            System.err.println("Log dosyası oluşturulamadı: " + e.getMessage());
        }
    }

    @Override
    public void log(TraceEvent event) {
        if (writer == null) return;

        try {
            //Convert event object to JSON string
            String jsonLine = objectMapper.writeValueAsString(event);
            
            //  Write the JSON line to the file and move to a new line (JSONL format).
            writer.write(jsonLine);
            writer.newLine();
            writer.flush(); //Every line is saved so that data is not lost even if the program crashes.
            
        } catch (IOException e) {
            System.err.println("Log yazma hatası: " + e.getMessage());
        }
    }

    @Override
    public void close() {
        try {
            if (writer != null) writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: JsonlTraceSink.py
import os
import json
import datetime
from tracer.trace_sink import TraceSink
from tracer.trace_event import TraceEvent

class JsonlTraceSink(TraceSink):
    """
    Trace olaylarını JSONL formatında dosyaya yazar.
    Her satır geçerli bir JSON objesidir.
    """

    def __init__(self):
        self.file = None
        self.setup_file()

    def setup_file(self) -> None:
        try:
            # logs klasörünü oluştur
            directory = "logs"
            if not os.path.exists(directory):
                os.makedirs(directory)

            # Dosya adını oluştur (run-YYYYMMDD-HHMMSS.jsonl)
            timestamp = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
            file_name = os.path.join(directory, f"run-{timestamp}.jsonl")

            # Writer'ı başlat (append modunda, utf-8 ile)
            # buffering=1 satır satır bufferlama yapar ama biz manuel flush da kullanacağız.
            self.file = open(file_name, "a", encoding="utf-8")
            
            print(f"Trace log started: {file_name}")

        except OSError as e:
            print(f"Log dosyası oluşturulamadı: {e}")

    def log(self, event: TraceEvent) -> None:
        if self.file is None:
            return

        try:
            # Event nesnesini dict'e çeviriyoruz (TraceEvent sınıfında to_dict yazmıştık)
            event_data = event.to_dict()
            
            # JSON string'e çevir. 
            # default=str: Eğer input/output içinde serileştirilemeyen (Enum gibi) nesneler varsa
            # onları string'e çevirerek hata almayı engeller.
            json_line = json.dumps(event_data, default=str)

            # Dosyaya yaz ve satır atla
            self.file.write(json_line + "\n")
            
            # Veri kaybını önlemek için diske yazmayı zorla (Flush)
            self.file.flush()

        except IOError as e:
            print(f"Log yazma hatası: {e}")

    def close(self) -> None:
        try:
            if self.file:
                self.file.close()
        except IOError as e:
            print(f"Dosya kapatma hatası: {e}")
